### PR TITLE
Fix dateByAddingDays

### PR DESCRIPTION
**Pull request by meypod**
Hi
It seems that dateByAddingDays does not work properly in some special timezones, on some special dates.
so I have edited it and added tests to make sure it does

### CODE REVIEW

**Comment by meypod** [comment]
I'm not sure if your use case is sensitive to time changes or not
but if it is, changing it from 1 hour in ms to 30 minutes should solve the issue, since the timezone changes and the daylight savings are usually rounded to 30 minutes or 1 hour

**Comment by meypod** [comment]
and actually I don't know if it makes any significant change at all, since this should be mostly a representation issue related to daylight savings

**Comment by z3bi** [comment]
@meypod so the tests you added 

```
const date1 = new Date(1667617200000);
         expect(date1.getDate()).toBe(5);
```

are constructing a date using the epoch value and with timezone-mock that ends up falling to a different calendar day based on that timezone's representation. We specifically avoid doing that so that all date calculations are based off of known calendar days in the UTC timezone. If we can find a test that fails using the `new Date(year, monthIndex, day)` constructor then we shouldn't definitely fix that.

File: test/math.test.ts [hunk 1]
```diff
@@ -101,10 +103,49 @@ test('rounding a date to the closest minute', () => {
...
 });
 
-test('adding days to date', () => {
-  const date1 = new Date(2015, 10, 1, 0, 0, 0);
-  expect(date1.getDate()).toBe(1);
+describe('test adding days to date on different time-zones', () => {
+  const testTimezone = (tz: string) =>
+    describe('timezone: ' + tz, () => {
+      beforeAll(() => {
+        timezoneMock.register('Brazil/East');
```
Comment (z3bi) [inline test/math.test.ts:110]: it looks like all these tests are in "Brazil/East" so why is the timezone string being passed in using the array below?

Comment (meypod) [reply test/math.test.ts]: you are right, I forgot to use tz, fixed

**Comment by z3bi** [comment]
Hmm so I think the problem here is actually `timezone-mock`. I recreated other tests for DST change over dates and set my local machine to that timezone and the tests passed.

**Comment by meypod** [comment]
To make sure it is not a quirk of mock timezone, try setting the time zone to the Brazil east and go to that special dates in tests with dst enabled, observe if going past 24:00 actually makes you go to the previous day or not
If it does, it is bound to happen

It doesn't sound illogical to me

**Comment by meypod** [comment]
But in the end, I'm not sure if this is anything significant, I'm doing some tests with my app tomorrow because I have observed some unusual stuff on my time zone on a few devices for sunnah times

It's probably a bug on my side, but in the case it was fixed by this patch, I'll let you know

**Comment by meypod** [comment]
Update: the patch is significant and fixed the issue on my faulty devices.

**Comment by z3bi** [comment]
> To make sure it is not a quirk of mock timezone, try setting the time zone to the Brazil east and go to that special dates in tests with dst enabled, observe if going past 24:00 actually makes you go to the previous day's 23:00 or not If it does, it is bound to happen
> 
> It doesn't sound illogical to me

This is part of the problem, Brazil no longer observes DST so this test doesn't actually make sense.

**Comment by z3bi** [comment]
@meypod if you remove `timezone-mock` we can add tests of different DST cutovers and then use https://github.com/marketplace/actions/set-timezone to run our tests on multiple timezones.

File: test/math.test.ts [hunk 2]
```diff
@@ -104,48 +102,47 @@ test('rounding a date to the closest minute', () => {
...
+    expect(date1.getDate()).toBe(4);
+
+    const date2 = dateByAddingDays(date1, 1);
+    expect(date2.getDate()).toBe(5);
+
+    const date3 = dateByAddingDays(date1, -1);
+    expect(date3.getDate()).toBe(3);
+  });
+
+  test('special date 3', () => {
```
Comment (z3bi) [inline test/math.test.ts:138]: please add comments to each of these special dates indicating which timezones they are testing

**Comment by meypod** [comment]
done
just as a note, the last special date I added is the date that was mis behaving in my app
on my faulty devices and tehran timezone, sunnah times were off by 3+ hours, plus the next day (March 21) was completely wrong (Fajr was around 12, instead of 4-5 am)

File: test/math.test.ts (hunk 2 above)
Comment (meypod) [reply test/math.test.ts:140]: added

**Comment by z3bi** [comment]
> done
> just as a note, the last special date I added is the date that was mis behaving in my app
> on my faulty devices and tehran timezone, sunnah times were off by 3+ hours, plus the next day (March 21) was completely wrong (Fajr was around 12, instead of 4-5 am)

I ran the new tests using the old dateByAddingDays function while my machine was in the Tehran timezone and the tests passed. This is not showing me a problem with the current approach.

File: test/math.test.ts (hunk 2 above)
Comment (z3bi) [reply test/math.test.ts:140]: Please use IANA timezone names, but also Brazil no longer does DST as I mentioned previously so I'm not sure what we're testing in this scenario. I believe this test only failed when using the incorrect timezone-mock library.

**Comment by meypod** [comment]
There's one issue remaining on my faulty devices
as of this year, Iran also doesn't see DST
this is no problem for newer devices
but on old android devices, DST seems to be causing the issue, hence applying this patch, fixes the weird times on those devices

But still, the time advances by 1 Hour on those devices, which is not ideal. I wish there was someway to disable DST ....

File: test/math.test.ts (hunk 2 above)
Comment (meypod) [reply test/math.test.ts:140]: well I dont know what is IANA names, I used the "TZ database name" column of the wiki

**Comment by meypod** [comment]
So overall, I have no idea how to test this tricky thing. I have seen it first hand that this patch does fix some issues in some cases.

at least the seemingly broken timezone mock would reproduce the issue and would test the issue

so idk, any suggestions ?

**Comment by z3bi** [comment]
I'm hesitant to change our date handling logic to try and patch a difficult to reproduce bug that only happens in a few older versions of Android. Additionally your change feels fragile to me with a while loop that is just checking to see if the string representation of the date is no longer the same.

If we can get a reproducible use case that doesn't require timezone-mock (which has incorrect DST data) then that's something we can fix.

**Comment by meypod** [comment]
well the devices were android 9 and 11, so idk if its that old

**Comment by meypod** [comment]
I agree the change seems fragile, but a better way did not cross my mind.

I can keep the changes local to my project, but I felt responsible to make you aware of this possible issue

**Comment by z3bi** [comment]
> well the devices were android 9 and 11, so idk if its that old

I was going off of your comment saying "old". It seems strange that basic Date math would be broken on Android 11's JavaScript implementation. Can you create the most barebones project that demonstrates this issue on Android and share the code?

**Comment by meypod** [comment]
I haven't try to reproduce on emulator, but my guess is I won't be able to reproduce, because of timezone and manufacture specific code. because why would a much older Samsung device on android 7 not face the issue, when 9 and 11 (of different brands) do.

I'll try to reproduce, but I wouldn't get my hopes up

File: test/math.test.ts (hunk 2 above)
Comment (z3bi) [reply test/math.test.ts:140]: In the "TZ Database name" modern timezones are in the format "Continent/City". "Brazil/East" is considered deprecated and instead "America/Sao_Paulo" is used.

**Comment by meypod** [comment]
hmm, it seems it's reproducible
I'll try to create a barebone react native project to demonstrate this (may take some time)

**Comment by meypod** [comment]
Okay
here is the reproduction repo (using expo)
https://github.com/meypod/adhan-js-bug/

**Comment by meypod** [comment]
I also made a Expo snack (for ease of use):
https://snack.expo.dev/@meypod/adhan-js-bug-snack?platform=android

**Comment by meypod** [comment]
you can use the android emulator provided by default by expo and it will have the bug, luckily:
image

as you can see, the first date shows times okay, but sunnah times, which are for the next day, have problem.

then there's second date, which is a mess, all of it)

then there's the third date, which is around 10 days before this, and all times are fine.

**Comment by meypod** [comment]
Also, I'm sorry, I have to correct myself
the samsung device is running android 9
the faulty devices are running android 10 and 11 (I know the one running 10 is a huawei honor phone)

File: test/math.test.ts (hunk 2 above)
Comment (sgtsquiggs) [reply test/math.test.ts:140]: these tests are not useful. `Date`'s constructor uses UTC timezone when you pass in a timestamp, which may be the inconsistency you are seeing.

Comment (meypod) [reply test/math.test.ts:140]: the problem is that it happens in a special environment, so probably you won't be able to test it in a normal environment

Comment (meypod) [reply test/math.test.ts:140]: at least with timezone-mock it was somewhat possible to test this

**Comment by meypod** [comment]
nice

**Pull request closed without merging**